allennlp/modules/seq2seq_encoders/efficient_multi_head_attention.py

```diff
@@ -0,0 +1,159 @@
+        factor applied to non-dropped elements.
+    """
+    def __init__(self,
+                 num_heads: int,
+                 input_dim: int,
+                 attention_dim: int,
+                 values_dim: int,
+                 output_projection_dim: int = None,
+                 attention_dropout_prob: float = 0.1) -> None:
+        super(EfficientMultiHeadAttention, self).__init__()
+
+        self._num_heads = num_heads
+        self._input_dim = input_dim
+        self._output_dim = output_projection_dim or input_dim
+        self._attention_dim = attention_dim
+        self._values_dim = values_dim
+
+        self._query_projections = Parameter(torch.FloatTensor(num_heads, input_dim, attention_dim))
+        self._key_projections = Parameter(torch.FloatTensor(num_heads, input_dim, attention_dim))
+        self._value_projections = Parameter(torch.FloatTensor(num_heads, input_dim, values_dim))
+
+        self._scale = input_dim ** 0.5
+        self._output_projection = Linear(num_heads * values_dim,
+                                         self._output_dim)
+        self._attention_dropout = Dropout(attention_dropout_prob)
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        # Because we are doing so many torch.bmm calls, which is fast but unstable,
+        # it is critically important to intitialise the parameters correctly such
+        # that these matrix multiplications are well conditioned initially.
+        # Without this initialisation, this (non-deterministically) produces
+        # NaNs and overflows.
+        init.xavier_normal(self._query_projections)
+        init.xavier_normal(self._key_projections)
+        init.xavier_normal(self._value_projections)
+
+    def get_input_dim(self):
+        return self._input_dim
+
+    def get_output_dim(self):
+        return self._output_dim
+
+    def forward(self,  # pylint: disable=arguments-differ
+                inputs: torch.Tensor,
+                mask: torch.LongTensor = None) -> torch.FloatTensor:
+        """
+        Parameters
+        ----------
+        inputs : ``torch.FloatTensor``, required.
+            A tensor of shape (batch_size, timesteps, input_dim)
+        mask : ``torch.FloatTensor``, optional (default = None).
+            A tensor of shape (batch_size, timesteps).
+
+        Returns
+        -------
+        A tensor of shape (batch_size, timesteps, output_projection_dim),
+        where output_projection_dim = input_dim by default.
+        """
+        num_heads = self._num_heads
+
+        batch_size, timesteps, hidden_dim = inputs.size()
+        if mask is None:
+            mask = Variable(inputs.data.new(batch_size, timesteps).fill_(1.0))
+
+        # Treat the queries, keys and values each as a ``num_heads`` size batch.
+        # shape (num_heads, batch_size * timesteps, hidden_dim)
+        inputs_per_head = inputs.repeat(num_heads, 1, 1).view(num_heads,
```

Comment:
`repeat` copies data, while `expand` only provides a view.  Do you really need `repeat` here?

Reply:
You're right, but I found the following thing which terrifies me a bit with `expand`, try this:
```
x = torch.FloatTensor([2])
x.repeat(1,3).mul_(2) # == [4, 4, 4]
x.expand(1,3).mul_(2) # == [16, 16, 16] !!
```
This is apparently the expected behaviour and a relic of how Torch used to work. I didn't know this and it makes me want to remove all inplace operations in the entire library.

Reply:
Wow, that is really scary.  But we don't do any in place operations at all except for `fill_`, do we?  They break autograd.

Reply:
I think pytorch expands and correctly tracks the operations actually, but i'll change this here.

Reply:
Actually, this has to be a `repeat` (or it's at least equivalent equivalent), because you can only `view` contiguous tensors, which expanded things aren't. Or else you end up having to call `.contiguous()` which I assume requires just as much memory as actually allocating memory initially.

Reply:
I wondered originally if that was why you wrote it that way.  Now we know =).  Thanks for checking.